CompareHands flow

daelynj opened the pull request:
The `CompareHands` class compares two hands by retrieving the value of each hand from the `Value` class, comparing them, and returning the winner.

The `Value` class holds scaffolding in this PR to demonstrate the flow of the code.

daelynj committed:
Add attr_reader to Card

Add the ability to access a Card objects rank and suit exclusively.

daelynj committed:
Add Value Class

Add Value Class to determine the score of a poker hand and return that
value.

daelynj committed:
Add retreive_value method

Add retreive_value method to determine the value of a poker hand using
the Value class.

daelynj committed:
Add winner method

Add winner method to return the winner of the 'poker game' based on who
has the best hand.

daelynj committed:
game.rb consistency

Add a comparison to determine_winner to remain consistent with changes
to how CompareHands determines the winner.

jgayfer commented on lib/poker_hands/CompareHands.rb:
```diff
@@ -1,5 +1,24 @@
+              Value.new.pair?(hand),
+              Value.new.high_card?(hand)
+            ]
+    value.max
```
We could skip the variable assignment here if we wanted
```
            [
              Value.new.royal_flush?(hand),
              Value.new.straight_flush?(hand),
              Value.new.four_of_a_kind?(hand),
              Value.new.full_house?(hand),
              Value.new.flush?(hand),
              Value.new.straight?(hand),
              Value.new.three_of_a_kind?(hand),
              Value.new.two_pair?(hand),
              Value.new.pair?(hand),
              Value.new.high_card?(hand)
            ].max
```

daelynj replied:
I thought it might be better to be explicit by writing `value.max` because the `.max` as you have it written could very easily be glanced over.

jgayfer commented on lib/poker_hands/value.rb:
```diff
@@ -0,0 +1,41 @@
+class Value
+  def royal_flush?(hand)
+    10
```
A method that ends in a question mark should generally return a boolean.

daelynj committed:
Update value methods to convention

By convention, methods that end in question marks return a boolean.

daelynj committed:
Update methods to maintain convention

By convention, methods that end in a question mark return a boolean.
To adhere to this, the value array is indexed such that the higher
scoring poker hands are closer to 0 than the lower scording hands. This
allows us to find the best score in a hand by returning the index of the
first true return from the Value class.

In the winner method we now return the hand that has the lowest value.
